pkg/api/persistentvolumeclaim/util.go

```diff
@@ -33,6 +34,61 @@ func DropDisabledFields(pvcSpec, oldPVCSpec *core.PersistentVolumeClaimSpec) {
 	if !dataSourceIsEnabled(pvcSpec) && !dataSourceInUse(oldPVCSpec) {
 		pvcSpec.DataSource = nil
 	}
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.RecoverExpansionFailure) && !allocatedResourcesInUse(oldPVCSpec) {
+		pvcSpec.AllocatedResources = nil
+	}
+}
+
+// SetAllocatedResources ensures that AllocatedResources field can only increase and tracks
+// maximum user requested capacity.
+func SetAllocatedResources(pvcSpec, oldPVCSpec *core.PersistentVolumeClaimSpec) {
+	if utilfeature.DefaultFeatureGate.Enabled(features.RecoverExpansionFailure) {
+		userResources := pvcSpec.Resources
+
+		// for new PVC creation we will simply set it to user requested size
+		// even if user provided one.
+		if oldPVCSpec == nil {
+			pvcSpec.AllocatedResources = &userResources
+			return
+		}
+
+		var oldAllocatedQuantity *resource.Quantity
+
+		if oldPVCSpec.AllocatedResources != nil {
+			oldAllocatedQuantity = oldPVCSpec.AllocatedResources.Requests.Storage()
+		}
+
+		// We generally do not want users to supply their own value of allocated resources
+		// but if old PVC did not had AllocatedResources set then we can use user requested
+		// resource.
+		if oldAllocatedQuantity == nil {
+			pvcSpec.AllocatedResources = &userResources
```

Comment:
okay, I have changed this code to default to `oldPVCSpec.Resources`  and it will be compared with `userResources` and whichever is higher will be used.